src/cachemachine/rubinimageinfo.py

```diff
@@ -0,0 +1,346 @@
+        ttag = tag[7:].title()
+        return f"Latest {ttag}"
+    if (
+        image_type == RubinImageType.RELEASE
+        or image_type == RubinImageType.RELEASE_CANDIDATE
+    ):
+        tagparts = tag.split("_")
+        rdigits = tagparts[0:3]
+        dtag = ".".join(rdigits)
+        if len(tagparts) > 3:
+            dtag += "-" + "_".join(tagparts[3:])
+    if image_type == RubinImageType.EXPERIMENTAL:
+        # Try running the stuff after exp_ through.  This is to handle the
+        #  common case of exp_<regular-tag>_something
+        etag = tag[4:]
+        etype = _type_from_tag(etag, recommended_tag)
+        if etype != RubinImageType.MALFORMED_TAG:
+            etname = etype.name.title()
+            cetag = _canonicalize_tag(etype, etag)
+            cetagparts = cetag.split("_")
+            etag = "_".join(cetagparts[1:])
+            dtag = f"{etname} {etag}"
+    if image_type == RubinImageType.MALFORMED_TAG:
+        dtag = tag  # Totally unparseable.
+    return f"{typename} {dtag}"
+
+
+@dataclass(frozen=True)
+class RubinImageInfo:
+    reference: str
+    """reference is a docker image reference.
+    See:
+     https://github.com/distribution/distribution/blob/main/reference/reference.go  # noqa: E501
+
+    Example: "registry.hub.docker.com/lsstsqre/sciplat-lab:w_2021_13@sha256:419c4b7e14603711b25fa9e0569460a753c4b2449fe275bb5f89743b01794a30"  # noqa: E501
+    """
+
+    host: str
+    """host is the host from which the image can be pulled.
+
+    Example: "registry.hub.docker.com"
+    """
+
+    repo: str
+    """image repository at a particular host.
+
+    Example: "lsstsqre"
+    """
+
+    name: str
+    """image name within a particular image repository.
+
+    example: "sciplat-lab"
+    """
+
+    tag: str
+    """image tag for a particular image.
+
+    Example: "w_2021_13"
+    """
+
+    digest: Optional[str]
+    """image digest for a particular image.
+
+    Example: "sha256:419c4b7e14603711b25fa9e0569460a753c4b2449fe275bb5f89743b01794a30"  # noqa: E501
+    """
+
+    reference_obj: reference.Reference
+    """field containing the results of parsing the reference string with the
+    docker_image class.
+    """
+
+    image_type: RubinImageType
+    """Rubin-specific image type
+
+    Example: RubinImageType.WEEKLY
+    """
+
+    semantic_version: semver.VersionInfo
+    """Semantic version constructed from a tag.  Only comparable between
+    RubinImageInfo instances of the same host, repo, name, and image_type.
+    """
+
+    display_name: str
+    """Human-readable display name corresponding to a tag."""
+
+    cycle: Optional[int]
+    """Cycle (used only in T&S SAL builds)."""
+
+    @classmethod
+    def from_reference(
+        cls,
+        ref: str,
+        recommended_tag: Optional[str] = None,
+        digest: Optional[str] = None,
+    ) -> "RubinImageInfo":
+        """Create a RubinImageInfo object from a reference string."""
```

Comment:
what is a reference string?

Reply:
https://github.com/distribution/distribution/tree/main/reference , specifically https://github.com/distribution/distribution/blob/main/reference/reference.go 

This is the not-quite-a-URL that identifies which Docker image you want to pull.  It is this thing that we do the initial parsing of with docker-image-py.